### PR TITLE
Implementation of TS clustering and tracking as envisioned in LUT firmware.

### PR DESCRIPTION
**Pull request by rodwyer100**
This push request takes the validated HLS firmware code and reincorporates it into the ldmx-sw structure. I have included in TrigScintFirmwareProducer a producer which takers clusterproducer_sw and trackproducer_sw (faithful reproductions of the code used to implement tracking in HLS), stages a collection of digis for insertion, and then obtains a collection of tracks for use thereafter. The producer as of yet does not save the collected digis into the track object, nor does it produces clusters, but it does faithfully reproduce the tracks of the existing horizontal geometry. There are two caveats with this:

1. It completely reproduces ldmx reconstruction results with no inefficiency in HLS because HLS does a better job at staging the rather poorly behaves ap_int classes it uses. I had some initial issues getting this code to be 100% efficient (its 98% efficient now) because ap_int behaves irregularly in how its implemented here.
2. This Firmware design was used to implement the original horiztonal 3 pad geometry (ldmx-det-v14) and does not have a LUT that corresponds to vertical bars. If this is desired, one could envision changing trackproducer_hw and testing it, but it will require further work.

- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.

I validated that the producer made the same number of tracks using print out statements like this:
10 10 10
18 18 18
29 29 30
 [ ElectronCounter ] 0 : Found 3 electrons (tracks) using input collection TriggerPadTracks_sim
I have not validated this code as extensively as the firmware version, so it could presumably make tracks from different clusters but my intuition about how the code works tells me that is very unlikely.

Here is a picture of the track number distribution for both:
Track no for original reconstruction:
image
Track no for firmware reconstruction:
image
If any more plots are necessary please lmk

### CODE REVIEW

**Comment by tvami** [comment]
> I am adding some includes in /includes/ which are necessary for the primitive non floating point objects that HLS uses to create firmware from C++ files. 

How are these files different from what we have in `Trigger/HLS_arbitrary_Precision_Types/include`? If they are not you should not add them again, but rather reuse them

File: TrigScint/exampleConfigs/firmwareEx.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+tsDigisDown.input_pass_name = "sim"
+
+tsClustersTag  =TrigScintClusterProducer.pad2()
+tsClustersUp  =TrigScintClusterProducer.pad1()
+tsClustersDown  =TrigScintClusterProducer.pad3()
+
+if "v12" in version :
+     tsClustersTag.pad_time = -2.
+     tsClustersUp.pad_time = 0.
+     tsClustersDown.pad_time = 0.
```
Comment (tvami) [inline TrigScint/exampleConfigs/firmwareEx.py]:
```suggestion
```

File: TrigScint/exampleConfigs/firmwareEx.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+from LDMX.TrigScint.trigScint import TrigScintDigiProducer
+from LDMX.TrigScint.trigScint import TrigScintClusterProducer
+from LDMX.TrigScint.trigScint import trigScintTrack
+from LDMX.TrigScint.trigScint import TrigScintFirmwareTracker
+
+if "v12" in version :
+     tsSimColls=[ "TriggerPadTagSimHits", "TriggerPadUpSimHits", "TriggerPadDnSimHits" ]
+else :
+     tsSimColls=[ "TriggerPad2SimHits", "TriggerPad3SimHits", "TriggerPad1SimHits" ]
```
Comment (tvami) [inline TrigScint/exampleConfigs/firmwareEx.py]: Is any of the v12 stuff still relevant?
```suggestion
tsSimColls=[ "TriggerPad2SimHits", "TriggerPad3SimHits", "TriggerPad1SimHits" ]
```

File: TrigScint/include/TrigScint/TrigScintFirmwareTracker.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  ldmx::TrigScintTrack makeTrack(Track outTrk);
+
+  /**
+   * add a hit at index idx to a cluster
+   */
+
+  void onProcessStart() override;
+
+  void onProcessEnd() override;
```
Comment (tvami) [inline TrigScint/include/TrigScint/TrigScintFirmwareTracker.h]:
```suggestion
```

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  std::cout<<"I GOT HERE 2"<<std::endl;
+  // TODO remove this once verified that the noise overlap bug is gone
+
+  int occupied[50];
+  for(int i = 0; i<50;i++){
+    occupied[i]=-1;
+  }
+  int count=0;
+  for (const auto &digi : digis1_) {
+    std::cout<<"I AM LOOPING IN DIGIS1 "<<digi.getPE()<<" "<<digi.getTime()<<std::endl;
```
Comment (tvami) [inline TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: Please move all the `cout`s to `ldmx_log` (probably debug level)
There are other `cout`s elsewhere so my comment here is rather general

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+  return;
+}
+
+void TrigScintFirmwareTracker::onProcessEnd() {
+  ldmx_log(debug) << "Process ends!";
+
+  return;
+}
+
```
Comment (tvami) [inline TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]:
```suggestion
```

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+  //std::cout<<"I GOT HERE 1"<<std::endl;
+  const auto digis1_{
+     event.getCollection<ldmx::TrigScintHit>(digis1_collection_, passName_)};
+  const auto digis3_{
+     event.getCollection<ldmx::TrigScintHit>(digis2_collection_, passName_)};
+  const auto digis2_{
+     event.getCollection<ldmx::TrigScintHit>(digis3_collection_, passName_)};
+
+  std::cout<<"I GOT HERE 2"<<std::endl;
+  // TODO remove this once verified that the noise overlap bug is gone
```
Comment (tvami) [inline TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: Is this bug gone?

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+	HPad1[count].Amp=(ap_int<12>)(digi.getPE());
+	HPad1[count].Time=(ap_int<12>)(digi.getTime());
+	occupied[digi.getBarID()]=count;
+	count++;
+      }
+    }
+  }
+
+  std::cout<<"I GOT HERE 3"<<std::endl;
+  for(int i = 0; i<50;i++){
```
Comment (tvami) [inline TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: please make the 50 a const expression, or a configurable if it is expected to change.

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+	topSeed=Point3[i].Seed.bID;
+      }
+    }
+  }
+
+   if (verbose_) {
+    ldmx_log(debug) << "Got digi collection " << digis1_collection_ << "_"
+                    << passName_ << " with " << digis1_.size() << " entries ";
+  }
+  std::cout<<"I GOT HERE 5"<<std::endl;
```
Comment (tvami) [inline TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: There is a few of these, please delete them
```suggestion
```

File: TrigScint/src/TrigScint/clusterproducer_sw.cxx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+#include "TrigScint/clusterproducer.h"
+
+Cluster* clusterproducer_sw(Hit inHit[NHITS]){
+
+	ap_int<12> SEEDTHR = 30;
+	ap_int<12> CLUSTHR = 30;
+
+	ap_int<12> mapL1[NCHAN];
+
+	Cluster* outClus = new Cluster[NCLUS];
```
Comment (tvami) [inline TrigScint/src/TrigScint/clusterproducer_sw.cxx]: can this become a smart pointer instead? I dont see any `delete`'s anywhere, so I'm worried for a memory leak

File: TrigScint/src/TrigScint/trackproducer_hw.cxx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+
+void trackproducer_hw(Cluster Pad1[NTRK],Cluster Pad2[NCLUS],Cluster Pad3[NCLUS],Track outTrk[NTRK],ap_int<12> lookup[NCENT][COMBO][2]){
+	#pragma HLS ARRAY_PARTITION variable=Pad1 dim=0 complete
+	#pragma HLS ARRAY_PARTITION variable=Pad2 dim=0 complete
+	#pragma HLS ARRAY_PARTITION variable=Pad3 dim=0 complete
+	#pragma HLS ARRAY_PARTITION variable=outTrk dim=0 complete
+	#pragma HLS ARRAY_PARTITION variable=lookup dim=0 complete
+	#pragma HLS PIPELINE II=10
+	std::cout<<"HELLO 2 HERE 1"<<std::endl;
```
Comment (tvami) [inline TrigScint/src/TrigScint/trackproducer_hw.cxx]: This file needs to be cleaned up too with all the HELLOs
```suggestion
```

File: scripts/ldmx-env.sh [hunk 11]
```diff
@@ -198,7 +198,6 @@ if hash docker &> /dev/null; then
       -e LDMX_BASE \
       -e DISPLAY=${LDMX_CONTAINER_DISPLAY}:0 \
       $_envs \
-      -v /tmp/.X11-unix:/tmp/.X11-unix \
```
Comment (tvami) [inline scripts/ldmx-env.sh]: why was this changed needed?

**Review by tvami — CHANGES_REQUESTED** [review]
I have a few smaller comments, please generally clean up the files, and you'll also have to run `just format-cpp` to have the clang formatting. I can also run the other sanitizers tomorrow for more comments, or maybe Tom will catch those before I run it tomorrow.

File: scripts/ldmx-env.sh (hunk 11 above)
Comment (rodwyer100) [reply scripts/ldmx-env.sh]: This change is necessary for anyone running on s3df. I had to use it but I can remove it.

File: TrigScint/exampleConfigs/firmwareEx.py (hunk 2 above)
Comment (rodwyer100) [reply TrigScint/exampleConfigs/firmwareEx.py]: No. I adopted an old working script here. Any naming convention works for me.

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx (hunk 4 above)
Comment (rodwyer100) [reply TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: Done.

File: TrigScint/src/TrigScint/trackproducer_hw.cxx (hunk 10 above)
Comment (rodwyer100) [reply TrigScint/src/TrigScint/trackproducer_hw.cxx]: done.

File: TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx (hunk 7 above)
Comment (rodwyer100) [reply TrigScint/src/TrigScint/TrigScintFirmwareTracker.cxx]: done

**Comment by rodwyer100** [comment]
We should talk about how ap_int is implemented in the ldmx-sw framework. I know TS in the Trigger repository doesn't use it, and I suspect it had something to do with the <*> in the fixed point library requiring ap_int to be found as a standard directory (and therefore not being found). While there are a couple repositories in the Ecal Trigger repositories that use it I am not certain how Christian managed to work with the relative paths indicated by the use of quotation marks and still have it built with <*>. That doesn't seem to work for me, though Im happy to take suggestions. I have implemented a fix that should work for both of us, and that is that in my local copy I have all of the includes of the HLS objects switched (wherever relevant) from <> to "". The issue is that since these are contained in a submodule we don't have access to, I can push that version which would work for us both. I know Tom worked on dereferening TrigScint and I am wondering if that is called for here, or if there is something better one should do. Otherwise I am not certain how to proceed on that front.

File: TrigScint/src/TrigScint/clusterproducer_sw.cxx (hunk 9 above)
Comment (rodwyer100) [reply TrigScint/src/TrigScint/clusterproducer_sw.cxx]: I'll try to do that :). Gonna hit the sack for tonight though!

File: TrigScint/python/trigScint.py [hunk 12]
```diff
@@ -285,6 +285,42 @@ def __init__(self,name) :
...
+        self.input_pass_name="" #take any pass
+        self.output_collection="TriggerPadTracks"
+        self.verbosity = 0
+
+class TrigScintFirmwareTracker(ldmxcfg.Producer) :
+    """Configuration for the track producer from the Firmware Tracker"""
+    def __init__(self,name) :
+        super().__init__(name,'trigscint::TrigScintFirmwareTracker','TrigScint')
+        self.clustering_threshold=40.0
+        self.digis1_collection='trigScintDigisPad1'
```
Comment (bryngemark) [inline TrigScint/python/trigScint.py:315]: should these be the QIEDigis?

**Review by bryngemark — COMMENTED** [review]
I have a general structure/philosophy comment/question. First, are these meant to be used in the HLS implementation or in sw clustering/tracking for comparison to HLS? If the former, I wonder if we should put them in some separate directory (be it within `TrigScint` or `Trigger`, I don't know). If the latter, I'd like the producer naming to be more explicit. You do use one explicit name "`TrigScintFirmwareTracker`" and that model is great. Overall it would be nice if you follow e.g. the capitalization conventions we're used to.

I would also appreciate if you commented the code more, so the next person can follow what's going on.

File: scripts/ldmx-env.sh (hunk 11 above)
Comment (tvami) [reply scripts/ldmx-env.sh]: Ah but there is a trick you can do on s3df, which you'll only have to do once per iana node you are on (so maybe twice ever? I usually end up on  006): 
```
touch /tmp/.X11-unix
```
and that's it

**Comment by tvami** [comment]
> The issue is that since these are contained in a submodule we don't have access to,

I'm confused by this statement. Doesn't this mean you just forgot the "recursive" when cloning in `git clone --recursive git@github.com:LDMX-Software/ldmx-sw.git` ? It should really be available like that

File: scripts/ldmx-env.sh (hunk 11 above)
Comment (tomeichlersmith) [reply scripts/ldmx-env.sh]: (This is one of a few small usability issues that are resolved with `denv` as well so please look at moving your workflow to using that program instead of `scripts/ldmx-env.sh`. https://ldmx-software.github.io/developing/quick-switch.html)

**Comment by tomeichlersmith** [comment]
Christian got the relative path includes working by telling CMake where to find the necessary headers.

https://github.com/LDMX-Software/ldmx-sw/blob/77fffe17e87e4237d9aaa106d9c343fc7f805947/Trigger/Algo/CMakeLists.txt#L39

This type of thing would be necessary in TrigScint as well if the `ap_*` headers are of use. Without a line like this in the CMake, there would be no way for the compiler to figure out where the headers are and thus would emit a `unable to find ap_XXX.h` style compiler error.

**Comment by tomeichlersmith** [comment]
I did a quick test and this branch still compiles after the following changes.
```
git rm -r TrigScint/include/TrigScint/etc
```
```diff
diff --git a/TrigScint/CMakeLists.txt b/TrigScint/CMakeLists.txt
index 2d1a8473..05382878 100644
--- a/TrigScint/CMakeLists.txt
+++ b/TrigScint/CMakeLists.txt
@@ -44,6 +44,7 @@ setup_library(module TrigScint
               dependencies Framework::Framework Recon::Event DetDescr::DetDescr
                            Tools::Tools SimCore::Event
 )
+target_include_directories(TrigScint PUBLIC ../Trigger/HLS_arbitrary_Precision_Types/include)
 
 setup_python(package_name LDMX/TrigScint)
 
diff --git a/TrigScint/include/TrigScint/objdef.h b/TrigScint/include/TrigScint/objdef.h
index 230cf112..b9706db9 100755
--- a/TrigScint/include/TrigScint/objdef.h
+++ b/TrigScint/include/TrigScint/objdef.h
@@ -1,7 +1,7 @@
 #ifndef OBJDEF_H
 #define OBJDEF_H
 
-#include "../../../Trigger/HLS_arbitrary_Precision_Types/include/ap_int.h"
+#include "ap_int.h"
 #define NTIMES 6
 #define NHITS 25
 #define NCLUS 25
```

File: TrigScint/python/trigScint.py (hunk 12 above)
Comment (rodwyer100) [reply TrigScint/python/trigScint.py:315]: I will comment the code more thoroughly for sure :)! I am including this code for two reasons. The first is that Elizabeth needs the code for efficiency studies to reflect firmware stuff. The other reason is along the lines of having firmware directly reflect software, and since the firmware uses pattern matching so will the software. This processor does both clustering and tracking in one step, using clusterproducer_sw and trackerproducer_hw to do those steps respectively. These files are the raw ones that are the validated pieces of HLS code, the processor itself just stages the digis. If I wanted to include the hitmaker I could as well, in which case I guess the input would be adc and tdc stuff from QIEDecoder, but I haven't done that just yet.

**Comment by rodwyer100** [comment]
I have added every change (smart pointers, general adoption of existing capitalization structure, etc) with the exception of clang formatting. I am trying to get my clang formatter working, but s3df doesnt allow for sudo so I have to git clone it and make it and Im on hour 6 or 7 of that. So hopefully it will all be done tomorrow. Please lmk if the chosen way of aligning with the capitalization structure is sufficient!

**Comment by tvami** [comment]
@rodwyer100 after we merge https://github.com/LDMX-Software/ldmx-sw/pull/1464 the clang format will automatically be applied if it's too much work for you. (Tho I use s3df too and `just format-cpp` works really nicely w/o any sudo needs)

File: scripts/ldmx-env.sh (hunk 11 above)
Comment (tvami) [reply scripts/ldmx-env.sh]: @rodwyer100 please revert the changes in this file

**Comment by rodwyer100** [comment]
The files have been clang formatted. Lmk if you think it needs anything else :)!

**Comment by tvami** [comment]
> The files have been clang formatted. Lmk if you think it needs anything else :)!

The only thing missing from my review is https://github.com/LDMX-Software/ldmx-sw/pull/1461#discussion_r1764339136

**Comment by rodwyer100** [comment]
Whoopsie just saw your comment. Fixed that change now. Thanks so much!

**Review by tvami — APPROVED** [review]
It's good from my side now, one more thing to change is the PR description, now that's not in line with the changes since you didnt move the headers in there, please modify the text accordingly

File: TrigScint/include/TrigScint/Firmware/trackproducer.h [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+#include "objdef.h"
+
+void copyCluster1(Cluster One, Cluster Two);
+void copyCluster2(Cluster One, Cluster Two);
+void trackproducer_ref(Cluster Pad1[NTRK], Cluster Pad2[NCLUS],
+                       Cluster Pad3[NCLUS], Track outTrk[NTRK],
+                       ap_int<12> lookup[NCENT][COMBO][2]);
+void trackproducer_hw(Cluster Pad1[NTRK], Cluster Pad2[NCLUS],
+                      Cluster Pad3[NCLUS], Track outTrk[NTRK],
+                      ap_int<12> lookup[NCENT][COMBO][2]);
```
Comment (tomeichlersmith) [inline TrigScint/include/TrigScint/Firmware/trackproducer.h:13]: It feels weird that Pad1 would have length `NTRK` while the other `Cluster` arrays would have length `NCLUS`. Is this on purpose?

File: TrigScint/src/TrigScint/Firmware/clusterproducer_sw.cxx [hunk 14]
```diff
@@ -0,0 +1,77 @@
+#include <stdio.h>
+
+#include <iostream>
+
+#include "TrigScint/Firmware/clusterproducer.h"
+#include "TrigScint/Firmware/objdef.h"
+
+std::unique_ptr<Cluster[]> clusterproducer_sw(Hit inHit[NHITS]) {
```
Comment (tomeichlersmith) [inline TrigScint/src/TrigScint/Firmware/clusterproducer_sw.cxx]: What you want to use is `std::array<Cluster, NCLUS>` since you know the length of the array at compile time. This allows you to pass around C-style, fixed-length arrays in C++ without needing to have bare pointers or any `new`.

**Review by tomeichlersmith — CHANGES_REQUESTED** [review]
I have one technical question and a technical change request.

I haven't looked through to understand the algorithm, but that's what future testing and bug reports are for I suppose.

File: TrigScint/include/TrigScint/Firmware/trackproducer.h (hunk 13 above)
Comment (rodwyer100) [reply TrigScint/include/TrigScint/Firmware/trackproducer.h:13]: Yes. The NTRK should be able to be smaller ultimately because we dont have the bandwidth in firmware to read out every track if there are more than 5 or so in a 25 ns clock cycle. The way the firmware handles this is that it handles the first cluster a little differently throughout, allowing it to read in a smaller number of clusters there. The number of tracks is less than or equal to the number of pad1 clusters and therefore the track max is inbuilt by a pad1 cluster max.

Comment (rodwyer100) [reply TrigScint/include/TrigScint/Firmware/trackproducer.h:13]: I think its set to 10, which is way more clusters than we expect in the ldmx environment on a single layer but it is still handleable by our firmware. This processor will try to model the firmware precisely.

Comment (tomeichlersmith) [reply TrigScint/include/TrigScint/Firmware/trackproducer.h:13]: That makes sense, can you add a comment explaining this so future developers don't think its a typo and replace it?

**Comment by tvami** [comment]
(not directly connected to this PR, but I'm happy to see the bot is doing the clang update as expected, Rory dont be surprised for this extra commit in your branch)
Screenshot 2024-09-18 at 11 12 07

File: TrigScint/src/TrigScint/Firmware/clusterproducer_sw.cxx (hunk 14 above)
Comment (tomeichlersmith) [reply TrigScint/src/TrigScint/Firmware/clusterproducer_sw.cxx]: This is what I'm imagining: https://godbolt.org/z/zbYzxKPzc

You'll have to manually enforce the shortening of `NCLUS` to `NTRK` but I think that is beneficial since it points out to the user a design of the firmware that we are intentionally emulating in the software.

**Review by tomeichlersmith — APPROVED** [review]
Thank you :) looks good
